added rand function to FunctionParser

asartori86 opened the pull request.

bangerth commented:
Using `srand()` and `rand()` has several problems. First, it is not thread safe. Second, and maybe more awkwardly, calling these global functions has the disadvantages that
* You get different results on different systems
* If you add a call to `rand()` somewhere before you start evaluating your function object, you will get different results
* In cases where you may or may not evaluate this function from an input file, you will get a different sequence of random numbers downstream.

All of these make debugging programs unnecessarily hard. For this reason, I would much prefer if the function you add had a static random number generator from BOOST, i.e. three lines of the kind
```
    static boost::random::mt19937 rng;
    boost::random::uniform_real_distribution<> uniform_distribution(0,1);
    double number = uniform_distribution(rng);
```
This way, the random numbers generated here are independent of the random numbers generated elsewhere. It is trivial to make this thread-safe by adding a mutex around it.

bangerth commented on source/base/function_parser.cc:
```diff
@@ -176,6 +179,22 @@ namespace internal
+    if (uninitialized)
+      {
+        if (seed > 0)
+          srand((unsigned int)seed);
```
This makes no sense. If someone provides a positive seed, then you call `srand` every single time with the same argument, and you will also get the very same (not so random) number in the following call to `rand()`.

asartori86 commented:
@bangerth , thanks for your feedback. I implemented using `boost` as you suggested. I implemented two functions `rand()` and `rand_seed(number)` so that `rand()` returns random number with random seed. `rand_seed` can be used for reproducible tests. Does it work for you?

bangerth commented on source/base/function_parser.cc:
```diff
@@ -176,6 +179,27 @@ namespace internal
+  {
+    boost::random::mt19937 rng;
+    rng.seed(static_cast<unsigned long>(std::time(0)));
+    boost::random::uniform_real_distribution<> uniform_distribution(0,1);
```
if you make the `rng` static, then you don't need to seed it every time.

This function is also missing the mutex.

bangerth commented on source/base/function_parser.cc:
```diff
@@ -176,6 +179,27 @@ namespace internal
+    std::mutex mtx;
+    mtx.lock();
+    static boost::random::mt19937 rng((unsigned int)seed);
+    mtx.unlock();
```
You'll have to make the `mutex` static as well, or it is not useful (it needs to be shared between all copies of the function that may run at the same time, but right now you have it a local variable).

You also need to expand the locked region until after the point where you actually *use* the random number generator. The easiest way is to use a scoped lock.

Finally, use `Threads::ThreadMutex` to stay compatible for non-C++11 compilers.

bangerth commented:
I think I still don't get the point of the `rand_seed` function. How is it supposed to be used? If you always call it with the same argument, then it's always going to return the same value. If you call it with something like `rand_seed(int(x))`, then you get different values everytime, but the sequence of random numbers you get is correlated in entirely unpredictable ways with the function you have as argument. In other words, I simply don't understand how you intend to use this function in input files...

asartori86 commented:
@bangerth, I've upadated the code and squashed the commit. I also add you as the author of the patch, if you don't mind.

asartori86 commented:
I'm using the `rand` function as a random noise of initial function. Using `rand_seed` allows me to have an initial solution with a "random" noise which is always the same and allow to make me a benchmark between, for example, direct and iterative solver. Let me add a movie :)

asartori86 commented:
![cahn-hilliard](https://cloud.githubusercontent.com/assets/10710861/12928707/4ffb2460-cf6f-11e5-84c6-d4985511ea0d.gif)

asartori86 commented:
in my `.prm` file I have this entry
```
subsection Initial solution
  set Function constants  = 
  set Function expression = 0.63+0.02*(0.5-rand_seed(9)); 0
  set Variable names      = x,y,t
end
```

bangerth commented:
Nice movie :-) 

I think I finally understand how `rand_seed(9)` works (at least how it works now). It's because the random number generator in this function is `static`, and so the initialization with `9` happens only once. I don't understand how this worked before, but this is no longer of concern to me :-)

Of course, if you had two function expressions in your input file, one referencing `rand_seed(9)` and the other `rand_seed(10)`, then only one of them will actually win because there is only one random number generator, and it will be initialized by the first expression to ever evaluated. Is this what you intend?


There were still a number of comments I had that are not yet addressed, see above.

luca-heltai commented:
I've just discussed with Alberto, and there is a patch coming which addresses the remaining comments plus your last concern. :+1:

asartori86 commented:
@bangerth , I guess we are all set thanks to the help of @luca-heltai :)

bangerth commented on source/base/function_parser.cc:
```diff
@@ -176,6 +180,34 @@ namespace internal
+    static std::map<double, boost::random::mt19937> rng_map;
+
+    if (rng_map.find(seed) == rng_map.end())
+      rng_map[seed] = boost::random::mt19937(static_cast<unsigned int>(seed));
```
Ah, clever solution. Care to add a comment somewhere in this function what the purpose of the `map` is (namely, to create a separate random number generator for every unique choice of random seed)?

asartori86 replied:
it is copyrighted by @luca-heltai  :)

bangerth commented on tests/base/function_parser_09.cc:
```diff
@@ -0,0 +1,58 @@
+void test()
+{
+  //  eval("rand()"); //random seed
+  eval("rand_seed(10)");
```
Could you also add a test that evaluates the `rand()` function just to make sure that the call succeeds? It will have to discard the value since it will not be reproducible, but it could test for its value being between 0 and 1.

asartori86 commented:
sure I'll do them

asartori86 commented:
@bangerth done

bangerth commented:
Thanks. By the way, want to post the video on the deal.II facebook or google+ pages?

luca-heltai commented:
![cahn-hilliard](https://cloud.githubusercontent.com/assets/2919756/12960160/10642880-d03a-11e5-9199-7a26007feb3b.gif)
Here's a nicer one, done with Alberto's code... :+1:

bangerth commented:
Nice :-) C-H is such a nice example for producing cool movies. Feel free to post on the Facebook or Google+ pages!
